File: Difficulty: Medium/Design MinMax Queue/design-min-max-queue.cpp
#include <queue>
#include <deque>
using namespace std;

class SpecialQueue {

  public:
    queue<int> q;  // Main queue to store elements
    deque<int> minDeque;  // Deque to track minimum elements
    deque<int> maxDeque;  // Deque to track maximum elements

    void enqueue(int x) {
        // Insert element into the queue
        q.push(x);
        
        // Maintain minDeque: remove elements greater than x from back
        while (!minDeque.empty() && minDeque.back() > x) {
            minDeque.pop_back();
        }
        minDeque.push_back(x);
        
        // Maintain maxDeque: remove elements smaller than x from back
        while (!maxDeque.empty() && maxDeque.back() < x) {
            maxDeque.pop_back();
        }
        maxDeque.push_back(x);
    }

    void dequeue() {
        // Check if queue is empty
        if (q.empty()) return;
        
        // Get the front element about to be removed
        int frontElement = q.front();
        q.pop();
        
        // Remove from minDeque if it's the current minimum
        if (frontElement == minDeque.front()) {
            minDeque.pop_front();
        }
        
        // Remove from maxDeque if it's the current maximum
        if (frontElement == maxDeque.front()) {
            maxDeque.pop_front();
        }
    }

    int getFront() {
        // Get front element
        if (q.empty()) {
            throw runtime_error("Queue is empty");
        }
        return q.front();
    }

    int getMin() {
        // Get minimum element
        if (minDeque.empty()) {
            throw runtime_error("Queue is empty");
        }
        return minDeque.front();  // Front of minDeque is always current min
    }

    int getMax() {
        // Get maximum element
        if (maxDeque.empty()) {
            throw runtime_error("Queue is empty");
        }
        return maxDeque.front();  // Front of maxDeque is always current max
    }
};

File: Difficulty: Medium/Design MinMax Queue/design-min-max-queue.java
class SpecialQueue {
    Deque<Integer>dq=new ArrayDeque<>();
    Map<Integer,Integer>mp=new HashMap<>();
    PriorityQueue<Integer>max=new PriorityQueue<>((a,b)->b-a);
    PriorityQueue<Integer>min=new PriorityQueue<>();

    public SpecialQueue() {
        // Define Data Structures
    }

    public void enqueue(int x) {
        // Insert element into the queue
        dq.addLast(x);
        mp.put(x,mp.getOrDefault(x,0)+1);
        max.add(x);
        min.add(x);
    }

    public void dequeue() {
        // Remove element from the queue
        int x=dq.removeFirst();
        mp.put(x,mp.getOrDefault(x,0)-1);
        if(mp.get(x)==0)mp.remove(x);
    }

    public int getFront() {
        // Get front element
        return dq.peekFirst();
    }

    public int getMin() {
        // Get minimum element
        while(!min.isEmpty() && !mp.containsKey(min.peek()))
        {
            min.remove();
        }
        return min.peek();
    }

    public int getMax() {
        // Get maximum element
        while(!max.isEmpty() && !mp.containsKey(max.peek()))
        {
            max.remove();
        }
        return max.peek();
    }
}